docs/source/usage.rst

```diff
@@ -337,58 +390,112 @@ Input YAML file:
-    /home/omry
+    >>> conf.user.name
+    'omry'
+    >>> conf.user.home
+    '/home/omry'
 
 You can specify a default value to use in case the environment variable is not defined.
-The following example sets `12345` as the the default value for the `DB_PASSWORD` environment variable.
+The following example sets `abc123` as the the default value when `DB_PASSWORD` is not defined.
 
 .. doctest::
 
     >>> cfg = OmegaConf.create({
-    ...       'database': {'password': '${env:DB_PASSWORD,12345}'}
+    ...       'database': {'password': '${env:DB_PASSWORD,abc123}'}
     ... })
-    >>> print(cfg.database.password)
-    12345
-    >>> OmegaConf.clear_cache(cfg) # clear resolver cache
-    >>> os.environ["DB_PASSWORD"] = 'secret'
-    >>> print(cfg.database.password)
-    secret
+    >>> cfg.database.password
+    'abc123'
+
+Environment variables are parsed when they are recognized as valid quantities that
+may be evaluated (e.g., int, float, dict, list):
+
+.. doctest::
+
+    >>> cfg = OmegaConf.create({
+    ...       'database': {'password': '${env:DB_PASSWORD,abc123}',
+    ...                    'user': 'someuser',
+    ...                    'port': '${env:DB_PORT,3306}',
+    ...                    'nodes': '${env:DB_NODES,[]}'}
+    ... })
+    >>> os.environ["DB_PORT"] = '3308'
+    >>> cfg.database.port  # converted to int
+    3308
+    >>> os.environ["DB_NODES"] = '[host1, host2, host3]'
+    >>> cfg.database.nodes  # converted to list
+    ['host1', 'host2', 'host3']
+    >>> os.environ["DB_PASSWORD"] = 'a%#@~{}$*&^?/<'
+    >>> cfg.database.password  # kept as a string
+    'a%#@~{}$*&^?/<'
+
 
 Custom interpolations
 ^^^^^^^^^^^^^^^^^^^^^
+
 You can add additional interpolation types using custom resolvers.
-This example creates a resolver that adds 10 the the given value.
+The example below creates a resolver that adds 10 to the given value.
 
 .. doctest::
 
-    >>> OmegaConf.register_resolver("plus_10", lambda x: int(x) + 10)
+    >>> OmegaConf.new_register_resolver("plus_10", lambda x: x + 10)
     >>> c = OmegaConf.create({'key': '${plus_10:990}'})
     >>> c.key
     1000
 
 
 Custom resolvers support variadic argument lists in the form of a comma separated list of zero or more values.
 Whitespaces are stripped from both ends of each value ("foo,bar" is the same as "foo, bar ").
-You can use literal commas and spaces anywhere by escaping (:code:`\,` and :code:`\ `).
+You can use literal commas and spaces anywhere by escaping (:code:`\,` and :code:`\ `), or
+simply use quotes to bypass character limitations in strings.
 
 .. doctest::
 
-    >>> OmegaConf.register_resolver("concat", lambda x,y: x+y)
+    >>> OmegaConf.new_register_resolver("concat", lambda x, y: x+y)
     >>> c = OmegaConf.create({
     ...     'key1': '${concat:Hello,World}',
     ...     'key_trimmed': '${concat:Hello , World}',
     ...     'escape_whitespace': '${concat:Hello,\ World}',
+    ...     'quoted': '${concat:"Hello,", " World"}',
```

Comment:
Changed in b1a3916 (I have a slight preference for yaml when the config gets larger, because I find it more readable, but it's not a big deal => I switched everything to dicts)